docs: tw919-batch-poster-recovery-and-troubleshooting (#3415)
OffchainLabs/arbitrum-docs: docs/launch-arbitrum-chain/operate/batch-poster-troubleshooting.mdx

<VanillaAdmonition type="info" title="Mempool size limits">

The exact mempool weight limits depend on the parent chain node's configuration, not the batch poster itself. The batch poster does not control the parent chain node's mempool size. If you are running your own parent chain node, consult its documentation for mempool configuration. If using a third-party RPC provider, the mempool limits are determined by the provider's infrastructure.

</VanillaAdmonition>

#### `ErrExceedsMaxMempoolSize` → "error posting batch"

- [DEBUG→WARN→ERROR over 5 min]—The next batch's nonce would exceed `unconfirmed nonce + max-mempool-transactions`, so it's held back.

> - [CAUSE]—L1 is confirming the poster's transactions slowly, and the in-flight window (default: 18) is filled.

Reviewer: ```suggestion
- [CAUSE]—<a data-quicklook-from="L1">L1</a> is confirming the poster's transactions slowly, and the in-flight window (default: 18) is filled.
```

- [MEANING]—Normal back-pressure short-term; sustained means posts are stuck.
- [ACTION]—If persistent, check L1 confirmation/fees, consider raising `max-mempool-transactions`.

#### `lack of L1 balance prevents posting transaction with desired fee cap`

- [WARN]—The poster's wallet can't cover the target max cost, so the bid is capped at the available balance.
- [CAUSE]—Low parent-chain balance and/or high L1 fees.
- [ACTION]—Fund the L1 wallet (`send-l1`); watch `arb/batchposter/wallet/eth`.

#### `a large batch posting backlog exists`

> - [INFO/WARN/ERROR]—Unposted messages are piling up; `INFO`: is recently hit L1 bounds, `WARN` at backlog > 10, `ERROR` at > 30.

Reviewer: ```suggestion
- [INFO/WARN/ERROR]—Unposted messages are piling up; `INFO` if it recently hit L1 bounds, `WARN` at backlog > 10, `ERROR` at > 30.
```

3. **Check parent chain connectivity**: Ensure the batch poster can reliably reach the parent chain RPC. Intermittent connectivity causes retry failures.
4. **Inspect RBF fee escalation**: If the error mentions "replacement transaction underpriced", the fee increase between attempts may be insufficient. The parent chain typically requires at least a 10% fee increase for RBF. Review the `blob-tx-replacement-times` schedule—shorter intervals may not allow enough price movement between attempts.

## Reverts and halting

#### `Large gap between last seen and current block number, skipping check for reverts`

- [WARN]—`pollForReverts` fell > 100 blocks behind the chain, so it skipped scanning the intervening blocks for reverted poster transactions and fast-forwarded.
- [CAUSE]—Node lag, a long pause, or a header-subscription stall.
- [MEANING]—Reverts in that skipped window won't be detected.
- [ACTION]—Usually transient; if frequent, investigate node sync/L1 RPC health.

#### `Transaction from batch poster reverted`

- [WARN→ERROR]—A batch posting transaction got a failed receipt on L1.

> - [ERROR (not warn)]—When using persistent storage, it sets `batchReverted` and **halts** all further posting.

Reviewer: ```suggestion
- [ERROR (not WARN)]—When using persistent storage, it sets `batchReverted` and **halts** all further posting.
```

- [CAUSE]—Contract rejected the batch (e.g., message-count mismatch, bad sequence number, gas/blob issue).

> - [ACTION]—Investigate `txErr`; restart clears the halt flag; a count mismatch after force-inclusion may need `allow-posting-first-batch-when-sequencer-message-count-mismatch`.

Reviewer: ```suggestion
- [ACTION]—Investigate `txErr`; restart clears the halt flag; a count mismatch after <a data-quicklook-from="forceinclusion">force-inclusion</a> may need `allow-posting-first-batch-when-sequencer-message-count-mismatch`.
```

#### `Error checking batch reverts`

- [DEBUG/WARN]—Revert check failed.

> - [DEBUG]—If the error contains "not found" (a benign parent-node inconsistency where one node served a header, another lacks), else WARN.

Reviewer: ```suggestion
- [DEBUG]—If the error contains "not found" (a benign parent-node inconsistency where one node served a header, another lacks it), else WARN.
```

- [ACTION]—Ignore the DEBUG case; investigate persistent WARNs.

## Nonce/sync

#### `failed to update nonce with queue empty; falling back to using a recent block`

- [WARN]—Couldn't get the finalized nonce, so it used a recent block instead. Safe because the queue is empty.
- [CAUSE]—Finality data unavailable/RPC issue.
- [ACTION]—Benign one-off; investigate if constant.

#### `Failed to get current nonce`

- [WARN]—Nonce refresh failed, but a previous nonce exists, so it's non-fatal.
- [CAUSE]—L1 RPC.
- [ACTION]—Ignore if transient.

#### `Failed to get latest nonce`

> - [WARN]—Couldn't fetch the unconfirmed nonce this loop; The iteration backs off 10s (`minWait`).

Reviewer: ```suggestion
- [WARN]—Couldn't fetch the unconfirmed nonce this loop; the iteration backs off 10s (`minWait`).
```

- [INFO]—The computed fee cap can't cover the current L1 base fee/required cost.
- [CAUSE]—L1 fees spiked above the poster's escalation target, or balanced-capped.
- [ACTION]—If posts stall, tune the fee formula (`target-price-gwei`, `urgency-gwei`, `max-fee-bid-multiple-bips`) or fund the wallet.

#### `submitting transaction with GasFeeCap less than latest basefee` / `...BlobGasFeeCap less than latest blobfee`

- [INFO]—Posting anyway with a cap below the current fee, expecting it to confirm as fees drop.
- [ACTION]—Normal during fee volatility; concerning only if posts never confirm.

#### `unable to fetch suggestedTipCap from l1 client to update arb/batchposter/suggestedtipcap metric`

- [WARN]—Couldn't get a tip suggestion for the metric.
- [MEANING]—Metric gap only.
- [ACTION]—Ignore if transient.

## L1 bounds/reorg

#### `Disabling batch posting due to batch being within reorg resistance margin from layer 1 minimum block or timestamp bounds`

- [ERROR]—The batch's first message is within `reorg-resistance-margin` of the L1 minimum bound, so posting is refused this round.

> - [CAUSE]—The margin guard (default 10m) protects against reorgs near the lower bound.

Reviewer: ```suggestion
- [CAUSE]—The margin guard (default 10m) protects against <a data-quicklook-from="reorg">reorgs</a> near the lower bound.
```

- [ACTION]—Expected safety behavior; set margin to `0` only if you accept the reorg risk.

#### `disabling L1 bound as batch posting message is close to the maximum delay`

- [ERROR]—Overriding the L1 block bound because messages are near `max-delay`; the `l1-block-bound-bypass` margin kicked in to avoid stalling.
- [ACTION]—Informational; means it chose to post over respecting the bounds.

#### `not posting more messages because block number or timestamp exceed L1 bounds`

- [INFO]—Stopped adding messages that fall outside the current L1 bound window.
- [ACTION]—Normal bounding behavior.

#### `error getting max time variation on L1 bound block; falling back on latest block`

> The L1 node couldn't give state at the finalized bound block, so it falls back to the latest block — usually transient.

Author: ```suggestion
The L1 node couldn't give state at the finalized bound block, so it falls back to the latest block—usually transient.
```

> #### `unknown L1 block bound config value; falling back on using finalized`.

Reviewer: ```suggestion
#### `unknown L1 block bound config value; falling back on using finalized`
```

- [WARN]—Couldn't get a tip suggestion for the metric.
- [MEANING]—Metric gap only.
- [ACTION]—Ignore if transient.

## L1 bounds/reorg

#### `Disabling batch posting due to batch being within reorg resistance margin from layer 1 minimum block or timestamp bounds`

- [ERROR]—The batch's first message is within `reorg-resistance-margin` of the L1 minimum bound, so posting is refused this round.
- [CAUSE]—The margin guard (default 10m) protects against reorgs near the lower bound.
- [ACTION]—Expected safety behavior; set margin to `0` only if you accept the reorg risk.

#### `disabling L1 bound as batch posting message is close to the maximum delay`

- [ERROR]—Overriding the L1 block bound because messages are near `max-delay`; the `l1-block-bound-bypass` margin kicked in to avoid stalling.
- [ACTION]—Informational; means it chose to post over respecting the bounds.

#### `not posting more messages because block number or timestamp exceed L1 bounds`

- [INFO]—Stopped adding messages that fall outside the current L1 bound window.
- [ACTION]—Normal bounding behavior.

#### `error getting max time variation on L1 bound block; falling back on latest block`

> - [WARN]—`unknown L2 block bound config value; falling back on using finalized`

Reviewer 2: ```suggestion
- [ERROR]—`unknown L1 block bound config value; falling back on using finalized`
```

> - [ERROR]—`unknown L1 block bound config value; falling back on using finalized`

Author: ```suggestion
- [WARN]—`unknown L1 block bound config value; falling back on using finalized`
```

- [WARN]—Redis lock check failed; it optimistically tries anyway.
- [CAUSE]—Redis connectivity.
- [ACTION]—Check Redis if high availability matters.

#### `Not posting batches right now because another batch poster has the lock or this node is behind`

- [DEBUG]—Normal on backup posters/when behind.
- [ACTION]—Expected; not an error.

#### `Failed to estimate gas for EIP-7623 check 1/2`

- [WARN]—An EIP-7623 calldata-cost estimation probe failed.
- [ACTION]—Usually transient; relevant only on EIP-7623 parent chains.

#### `error estimating gas for batch`

- [escalates via ephemeral handler]—Gas estimation failed (`ErrNormalGasEstimationFailed`); DEBUG→WARN→ERROR over 5 min.
- [CAUSE]—L1 state lag, reverting estimation, inbox not caught up.
- [ACTION]—Investigate if it reaches `ERROR`.

## Config/startup

#### `max-size is deprecated; use max-calldata-batch-size...`

- [ERROR]—Deprecated flag in use.
- [ACTION]—Migrate to `max-calldata-batch-size`.

#### `Disabling data poster storage, as parent chain appears to be an Arbitrum chain without a mempool`

- [INFO]—Auto-switched to no-op storage on an Arbitrum parent.
- [ACTION]—Expected for L3s.

> #### `messagesPerbatch is somehow zero`

Reviewer 2: ```suggestion
#### `messagesPerBatch is somehow zero`
```

- [WARN]—Couldn't get a tip suggestion for the metric.
- [MEANING]—Metric gap only.
- [ACTION]—Ignore if transient.

## L1 bounds/reorg

#### `Disabling batch posting due to batch being within reorg resistance margin from layer 1 minimum block or timestamp bounds`

- [ERROR]—The batch's first message is within `reorg-resistance-margin` of the L1 minimum bound, so posting is refused this round.
- [CAUSE]—The margin guard (default 10m) protects against reorgs near the lower bound.
- [ACTION]—Expected safety behavior; set margin to `0` only if you accept the reorg risk.

#### `disabling L1 bound as batch posting message is close to the maximum delay`

- [ERROR]—Overriding the L1 block bound because messages are near `max-delay`; the `l1-block-bound-bypass` margin kicked in to avoid stalling.
- [ACTION]—Informational; means it chose to post over respecting the bounds.

#### `not posting more messages because block number or timestamp exceed L1 bounds`

- [INFO]—Stopped adding messages that fall outside the current L1 bound window.
- [ACTION]—Normal bounding behavior.

#### `error getting max time variation on L1 bound block; falling back on latest block`

> 

Author: ```suggestion
The L1 node couldn't give state at the finalized bound block, so it falls back to the latest block — usually transient.
```

> #### `unknown L1 block bound config value; falling back on using finalized`. The L1 node couldn't give state at the finalized bound block, so it falls back to the latest block — usually transient.

Author: ```suggestion
#### `unknown L1 block bound config value; falling back on using finalized`. 
```

- [WARN]—Couldn't get a tip suggestion for the metric.
- [MEANING]—Metric gap only.
- [ACTION]—Ignore if transient.

## L1 bounds/reorg

#### `Disabling batch posting due to batch being within reorg resistance margin from layer 1 minimum block or timestamp bounds`

- [ERROR]—The batch's first message is within `reorg-resistance-margin` of the L1 minimum bound, so posting is refused this round.
- [CAUSE]—The margin guard (default 10m) protects against reorgs near the lower bound.
- [ACTION]—Expected safety behavior; set margin to `0` only if you accept the reorg risk.

#### `disabling L1 bound as batch posting message is close to the maximum delay`

- [ERROR]—Overriding the L1 block bound because messages are near `max-delay`; the `l1-block-bound-bypass` margin kicked in to avoid stalling.
- [ACTION]—Informational; means it chose to post over respecting the bounds.

#### `not posting more messages because block number or timestamp exceed L1 bounds`

- [INFO]—Stopped adding messages that fall outside the current L1 bound window.
- [ACTION]—Normal bounding behavior.

#### `error getting max time variation on L1 bound block; falling back on latest block`

> - [WARN]-The L1 node couldn't give state at the finalized bound block, so it falls back to the latest block — usually transient.

Author: ```suggestion
```
